[docs] Add a section to the authentication guide about exchanging access tokens (#15444)
expo/expo: docs/pages/guides/authentication.md

- Be sure to disable the prompt until `request` is defined.
- You can only invoke `promptAsync` in a user-interaction on web.

## Obtaining access tokens

The vast majority of providers described here implement the [OAuth 2](https://oauth.net/2/) standard for secure authentication/authorization.
In the authorization code grant, the identity provider returns a one-time code which is then exchanged for the user's access token. Because
[your application code is not a secure place to store secrets](https://reactnative.dev/docs/security#storing-sensitive-info), it is necessary to
exchange the authorization code in a context (e.g., your server) where you can securely store and use a client secret to access the provider's

> token endpoint.

Author: This is true for OAuth integrations such as GitHub that only allow clients to be configured as confidential - that is, not public. Public clients can of course exchange authorization codes (preferably using PKCE) but many providers hide this implementation detail from users registering clients. E.g., GitHub's documentation [shows the client secret as required](https://docs.github.com/en/developers/apps/building-oauth-apps/authorizing-oauth-apps#2-users-are-redirected-back-to-your-site-by-github) when exchanging a code for an access token.

Author: I'm not sure how in the weeds we need to make this; the inline docs talk about needing to do the code exchange in a secure environment, so this PR is just to call this out more clearly right at the top of the guide.

> The vast majority of providers described here implement the [OAuth 2](https://oauth.net/2/) standard for secure authentication/authorization.
> In the authorization code grant, the identity provider returns a one-time code which is then exchanged for the user's access token. Because
> [your application code is not a secure place to store secrets](https://reactnative.dev/docs/security#storing-sensitive-info), it is necessary to
> exchange the authorization code in a context (e.g., your server) where you can securely store and use a client secret to access the provider's
> token endpoint.

Reviewer: ```suggestion
The vast majority of providers described here implement the [OAuth 2](https://oauth.net/2/) standard for secure authentication/authorization. In the authorization code grant, the identity provider returns a one-time code, which is exchanged for the user's access token. Since [your application code is not a secure place to store secrets](https://reactnative.dev/docs/security#storing-sensitive-info), it is necessary to exchange the authorization code in a context (for example, your server) where you can securely store and use a client secret to access the provider's token endpoint.
```
